src/main/java/com/example/Ecommerce/service/VoucherServiceImpl.java

```diff
@@ -0,0 +1,128 @@
+
+@Service
+@RequiredArgsConstructor
+public class VoucherServiceImpl implements VoucherService {
+
+    private final VoucherRepo voucherRepo;
+    private final ProductRepo productRepo;
+
+
+    @Override
+    public VoucherDto createVoucher(VoucherDto voucherDto) {
+
+        if (voucherRepo.findByCode(voucherDto.getCode()) != null) {
+            throw new ResourceNotFoundException("Voucher with code already exists: " + voucherDto.getCode());
+        }
+        checkIfVoucherIsExpired(voucherDto.getExpiryDate());
+
+        Voucher voucher = VoucherMapper.convertDtoToEntity(voucherDto);
+        Voucher savedVoucher = voucherRepo.save(voucher);
+        return VoucherMapper.convertEntityToDto(savedVoucher);
+    }
+
+
+    @Override
+    public Voucher findVoucherByCode(String code) {
+
+        Voucher voucher = voucherRepo.findByCode(code);
+        if (voucher == null) {
+            throw new ResourceNotFoundException("Voucher not found with code: " + code);
+        }
+        return voucher;
+    }
+
+    @Override
+    public void deleteVoucher(Long id) {
+        checkIfVoucherExistsById(id);
+        updateProductsWithoutVoucher(id);
+        voucherRepo.deleteById(id);
+    }
+
+    @Override
+    public void applyVoucherDiscount(Product product) {
+
+        if (product.getVoucherCode() != null) {
+            Voucher voucher = product.getVoucherCode();
+            if (isVoucherDateValid(voucher.getExpiryDate())) {
+                BigDecimal discountedPrice = calculateDiscountedPrice(product.getPrice(), voucher.getDiscount());
+                product.setPrice(discountedPrice.doubleValue());
+            } else {
+                product.setPrice(product.getPrice());
+                deleteVoucher(voucher.getId());
+            }
+        }
+    }
+
+    @Override
+    @Transactional
+    public void addProductsToVoucher(Long voucherId, List<Long> productIds) {
+        checkIfVoucherExistsById(voucherId);
+        Voucher voucher = voucherRepo.findById(voucherId).orElse(null);
+
+        List<Product> products = productRepo.findAllById(productIds);
+
+        if (products.size() != productIds.size()) {
+            throw new ResourceNotFoundException("One or more product IDs do not match any existing products.");
+        }
+
+        for (Product product : products) {
+            product.setVoucherCode(voucher);
+        }
+
+        productRepo.saveAll(products);
+    }
+
+
+    private void checkIfVoucherExistsById(Long id) {
+        if (voucherRepo.findById(id).isEmpty()) {
+            throw new ResourceNotFoundException("Voucher not found with id: " + id);
+        }
+    }
+
+    private void checkIfVoucherIsExpired(Date expireDate) {
+        LocalDateTime now = LocalDateTime.now();
+        LocalDateTime expireDateTime = convertToLocalDateTime(expireDate);
+
+        if (expireDateTime.isBefore(now)) {
+            throw new ResourceNotFoundException("Voucher has expired.");
```

Comment:
I believe this Exception is not correct. `checkIfVoucherIsExpired` indicates that this function checks if it's expired or not, so it should return boolean, and the caller can then throw an exception if it's not true.

another option is to rename it to `assertIfVoucherIsExpired` and in this case it would throw exception like `VoucherExpiredException`